Source/Charts/Charts/BarLineChartViewBase.swift

```diff
@@ -816,7 +816,7 @@ open class BarLineChartViewBase: ChartViewBase, BarLineScatterCandleBubbleChartD
         
         matrix = _viewPortHandler.refresh(newMatrix: matrix, chart: self, invalidate: true)
         
-        if delegate !== nil
+        if matrix != originalMatrix
```

Comment:
@jjatie struct in swift 4.2 is still copied on assigning right?
`refresh` will override `_touchMatrix` with `matrix`

@anton-filimonov in `refresh`->`limitTransAndScale()`, so `matrix` will be different anyway and makes `matrix != originalMatrix` always true to me?

Reply:
@liuxuan30 yes, value types are copied on assigning, but that doesn't mean that comparing those values is useless because they are compared by value not by address. And since we are on the edge of available region we always get equal matrixes back from `refresh()->limitTransAndScale()`. And test shows that this change prevents calls of `chartTranslated` when no translation really happens.
The thing I really forgot to do is to do the same check when scale is performed. So I'll update this PR with that change soon

Reply:
@liuxuan30 well I checked the situation with scale and same change there is not needed this much because delegate method calls there are limited with some other conditions. False calls are still possible there but it's only one call, for example when we zoom out the chart when it's already zoomed out to minimum (on the first call object think that we zoom in because no scale was made before)
So I'm not sure this change is needed. What do you think about it?

Reply:
@jjatie 
>A value type is a type whose value is copied when it’s assigned to a variable or constant, or when it’s passed to a function.

from https://docs.swift.org/swift-book/LanguageGuide/ClassesAndStructures.html

Reply:
@anton-filimonov sorry I made a mistake that I thought this is a func inside ChartsTransformer. Actually it's in `performPanChange()`, which is a pan gesture handler..

and in this case, when would matrix equals to originalMatrix for you to prevent the delegate get called? 

I checked the issue you posted:
```
scroll to edge
try to scroll in direction of the same edge
```
you are trying to scroll, which means a delegate should be called?

Reply:
@liuxuan30 I'm trying to scroll, but no scroll happens because chart is on the edge. So chart is not translated, nothing changes and seems like `chartTranslated` shouldn't be called in such a case

Reply:
I see, so it's on the edge. But, if you are on the edge, would matrix equals to originalMatrix in this case? I just wonder they wouldn't be equal at any time.

Reply:
Yes, matrix stays unchanged in this case. Actually we get new matrix, created in `limitTransAndScale` but that matrix is equal to previously used value

Reply:
I have tested on the edge, there are some `matrix == originalMatrix` cases, but there are also multiple false cases. from my understanding, only a few can be filtered out, is that what you are saying `So I'm not sure this change is needed.`?

Reply:
`So I'm not sure this change is needed.` was about adding the same change to pinch gesture handling before calling `chartScaled`. There are two checks before all the work is done (`canZoomMoreX`, `canZoomMoreY`) and they eliminate almost all false calls, that's why I'm not sure this change is needed for that case.

Reply:
It's much easier to reproduce the issue with false calls of `chartTranslated` when you scroll to the corner (I edited two last steps to reproduce the issue in #3803 ) so that translation is blocked in two directions and every scrolling step will cause no change to the transformation matrix

Reply:
`but there are also multiple false cases mixing up at the same time` Probably some translation still occurs in this case. Maybe for example you reached the right edge but scrolled right-up.

Reply:
Yes I mean, adding this fix can filter a few calls, but generally, it won't prevent the delegate get called no matter where you are. You cannot rely on it to judge if you are on edge or something.